src/opera.jl

```diff
@@ -1,72 +1,262 @@
 
-See [`opera_enqueue!`](@ref).
+See [`poke`](@ref) and [`@call`](@ref).
 """
 mutable struct Opera
-    calls::PriorityQueue{AbstractOperaCall, Float64}
+    # dictionary of `uuid => agent` pairs
     directory::Dict{UUID, AbstractAlgebraicAgent}
+    # intantious interactions
+    instantious_interactions::Vector{InstantiousInteraction}
+    # futures
+    futures::Vector{Future}
+    futures_log::Vector{FutureLog}
+    # controls
+    controls::Vector{Control}
+    controls_log::Vector{ControlLog}
 
     function Opera(uuid2agent_pairs...)
-        new(PriorityQueue{AbstractOperaCall, Float64}(Base.Order.Reverse),
-            Dict{UUID, AbstractAlgebraicAgent}(uuid2agent_pairs...))
+        new(Dict{UUID, AbstractAlgebraicAgent}(uuid2agent_pairs...),
+            Vector{InstantiousInteraction}(undef, 0),
+            Vector{Future}(undef, 0),
+            Vector{FutureLog}(undef, 0),
+            Vector{Control}(undef, 0),
+            Vector{ControlLog}(undef, 0))
+    end
+end
+
+# dispatch on the interaction call, and execute it
+function call(opera::Opera, call::Function)
+    if hasmethod(call, Tuple{})
+        call()
+    elseif hasmethod(call, Tuple{Opera})
+        call(opera)
+    elseif length(opera.directory) > 1 &&
+           hasmethod(call, Tuple{typeof(topmost(first(opera.directory).value))})
+        call(topmost(first(opera.directory).value))
+    else
+        @error """interaction $call must have one of the following forms:
+            - be parameterless,
+            - be a function of `Opera` instance,
+            - be a function of the topmost agent in the hierarchy.
+        """
     end
 end
 
 "Schedule an algebraic interaction."
-function opera_enqueue!(::Opera, ::AbstractOperaCall, ::Float64) end
+function add_instantious_interaction!(opera::Opera, action::InstantiousInteraction)
+    # sorted insert
+    pushfirst!(opera.instantious_interactions, action)
+
+    ix = 1
+    while ix < length(opera.instantious_interactions)
+        if action.priority > opera.instantious_interactions[ix + 1].priority
+            opera.instantious_interactions[ix] = opera.instantious_interactions[ix + 1]
+            opera.instantious_interactions[ix + 1] = action
+            ix += 1
+        else
+            break
+        end
+    end
+end
 
-function opera_enqueue!(opera::Opera, call::AgentCall, priority::Float64 = 0.0)
-    !haskey(opera.calls, call) && enqueue!(opera.calls, call => priority)
+# Execute instantious interactions
+function execute_instantious_interaction!(opera::Opera)
+    while !isempty(opera.instantious_interactions)
+        call(opera, pop!(opera.instantious_interactions).call)
+    end
+end
+
+"""
+    poke(agent, priority=0)
+Schedule an interaction. Interactions are implemented within an instance `Opera`, sorted by their priorities.
+Internally, reduces to `_interact!(agent)`.
+
+See also [`Opera`](@ref).
+
+# Examples
+```julia
+poke(agent, 1.)
+```
+"""
+function poke(agent, priority = 0.0)
```

Comment:
The difference is that 
 - `poke` just invokes `_interact(agent)`, and if we can proceed with the function form, I think it should be preferred to macro forms in general,
 - `@call ... expression` wraps `expression` into an anonymous function `() -> expression`, so that you may conveniently write `@call model wake_up_agent(my_agent)` - this is the motivation for the macro form.

Reply:
Understood, that makes sense. Maybe it would be nice to add that tiny example to the design.md doc so users are most clear about it.